src/pip/_internal/index.py

```diff
@@ -939,15 +939,14 @@ def _get_encoding_from_headers(headers):
     return None
 
 
-_CLEAN_LINK_RE = re.compile(r'[^a-z0-9$&+,/:;=?@.#%_\\|-]', re.I)
-
-
 def _clean_link(url):
     # type: (str) -> str
     """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
     the link, it will be rewritten to %20 (while not over-quoting
     % or other characters)."""
-    return _CLEAN_LINK_RE.sub(lambda match: '%%%2x' % ord(match.group(0)), url)
+    result = urllib_parse.urlparse(url)
+    return urllib_parse.urlunparse(
+        result._replace(path=urllib_parse.quote(result.path)))
```

Comment:
I believe you have a bug in your implementation. You should be calling `unquote()` on the path before calling `quote()`. Otherwise, you will have double-escaping. You should include a test case to check that you are not double-escaping. Also, maybe this could be a cause for the test failures, though I don't know why it would only occur on Windows.

Reply:
Done, thanks for pointing that out!